File: 2018/12/12.py
def parse(file):
    mappings = {}
    with open(file) as f:
        state = f.readline().strip().split(' ')[2]
        f.readline()
        for line in f.readlines():
            f, t = line.strip().split(' => ')
            mappings[f] = t
    return state, mappings


def iterate(state, mappings):
    state = {
        i: s for i, s in enumerate(state)
    }
    n = ''
    for i in range(len(state) + 2):
        a = ''.join(
            state.get(l, '.') for l in range(i-2, i+3)
        )
        n += mappings.get(a, '.')
    
    return n.rstrip('.')


s, m = parse('input')
s = '......' + s
for i in range(20):
    s = iterate(s, m)

pots = 0
for i, c in enumerate(s):
    if c == '#':
        pots += i - 6

def part_two(x):
    return 7236 + (x - 108) * 67

print('Part 1:', pots)
print('Part 2:', part_two(50000000000))

File: 2018/12/12.go
package main

import "fmt"
import "io/ioutil"
import "strings"

type Mapping map[string]string
type State map[int]string

func parseFile(fileName string) (State, Mapping) {
	state := make(State)
	for i := -5; i < 0; i++ {
		state[i] = "."
	}
	mapping := make(Mapping)
	data, _ := ioutil.ReadFile(fileName)
	for i, line := range strings.Split(string(data), "\n") {
		if len(line) == 0 {
			continue
		}
		if i == 0 {
			p := strings.Split(line, " ")
			for k, c := range p[2] {
				state[k] = string(c)
			}
		}
		if i > 1 {
			p := strings.Split(line, " ")
			mapping[p[0]] = p[2]
		}
	}
	return state, mapping
}

func extract(state State, i int) string {
	s := ""
	for j := -2; j < 3; j++ {
		v, ok := state[i+j]
		if !ok {
			v = "."
		}
		s += v
	}
	return s
}

func iterate(state State, mapping Mapping) State {
	newState := make(State)
	for i := -5; i < len(state)-4; i++ {
		neighbourhood := extract(state, i)
		newValue, ok := mapping[neighbourhood]
		if !ok {
			newValue = "."
		}
		newState[i] = newValue
	}
	return newState
}

func draw(state State, iteration int) {
	fmt.Print(fmt.Sprintf("%2d:", iteration))
	for i := -5; i < len(state); i++ {
		fmt.Print(state[i])
	}
	fmt.Println(" ", score(state))
}

func score(state State) int {
	s := 0
	for k, v := range state {
		if v == "#" {
			s += k
		}
	}
	return s
}

func lateScore(rounds int) int {
	return rounds * 67
}

func main() {
	state, mapping := parseFile("input")
	for i := 1; i < 21; i++ {
		state = iterate(state, mapping)
	}
	fmt.Println("Part 1:", score(state))
	fmt.Println("Part 2:", lateScore(50000000000))
}
